src/document-fetcher.js

```diff
@@ -0,0 +1,164 @@
+   * @param {!FetchInitDef} init
+   * @private
+   */
+  xhrRequest_(input, init) {
+    return new Promise((resolve, reject) => {
+      this.xhr_.open(init.method || 'GET', input, true);
+      if (init.credentials == 'include') {
+        this.xhr_.withCredentials = true;
+      }
+      this.xhr_.responseType = 'document';
+      if (init.headers) {
+        Object.keys(init.headers).forEach(header => {
+          this.xhr_.setRequestHeader(header, init.headers[header]);
+        });
+      }
+      this.xhr_.onreadystatechange = () => {
+        if (this.xhr_.readyState < /* STATUS_RECEIVED */ 2) {
+          return;
+        }
+        if (this.xhr_.status < 100 || this.xhr_.status > 599) {
+          this.xhr_.onreadystatechange = null;
+          reject(user().createExpectedError(
+              `Unknown HTTP status ${this.xhr_.status}`));
+          return;
+        }
+
+        // TODO(dvoytenko): This is currently simplified: we will wait for the
+        // whole document loading to complete. This is fine for the use cases
+        // we have now, but may need to be reimplemented later.
+        if (this.xhr_.readyState == /* COMPLETE */ 4) {
+          const options = {
+            status: this.xhr_.status,
+            statusText: this.xhr_.statusText,
+            headers: parseHeaders(this.xhr_.getAllResponseHeaders()),
+          };
+          options.url = 'responseURL' in this.xhr_
+            ? this.xhr_.responseURL : options.headers['X-Request-URL'];
+          const body = 'response' in this.xhr_
+            ? this.xhr_.response : this.xhr_.responseText;
+          resolve(new Response(/** @type {string} */ (body || ''), /** @type {!ResponseInit} */ (options)));
+        }
+      };
+      this.xhr_.onerror = () => {
+        reject(user().createExpectedError('Network failure'));
+      };
+      this.xhr_.onabort = () => {
+        reject(user().createExpectedError('Request aborted'));
+      };
+      if (init.method == 'POST') {
+        this.xhr_.send(/** @type {!FormData} */ (init.body));
+      } else {
+        this.xhr_.send();
+      }
+    });
+  }
+
+  /**
+   *
+   *
+   * @param {string} input
+   * @param {?FetchInitDef=} opt_init
+   * @return {!Promise<!Document>}
+   * @ignore
+   */
+  fetchDocument(input, opt_init) {
+    const init = setupInit(opt_init, 'text/html');
+    init.responseType = 'document';
+    return this.fetchAmpCors_(input, init).then(response => {
+      assertSuccess(/** @type {!Response} */(response));
+      if (!this.viewerResponded_) {
+        return this.xhr_.responseXML;
+      } else {
+        // Since this reponse is provided by viewer,
+        // doing a DOMParser on main thread is our only option.
+        return response.text().then(responseText =>
+          new DOMParser().parseFromString(responseText, 'text/html'));
+      }
+    });
+  }
+}
+
+/**
+ *
+ * @param {string} rawHeaders
+ * @return {JsonObject}
+ */
+function parseHeaders(rawHeaders) {
```

Comment:
We might be able to delete this.

Reply:
post an explanation above, of why we cant